### PR TITLE
Make CoAP responses carry DTLS context as well as requests

### PR DESCRIPTION
**Pull request by akolosov-n**

### CODE REVIEW

File: coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java [hunk 1]
```diff
@@ -15,24 +15,41 @@
...
+import java.util.function.BiFunction;
+import org.opencoap.ssl.netty.DatagramPacketWithContext;
 import org.opencoap.ssl.transport.DtlsSessionContext;
 
 public class DtlsTransportContext {
     public static final TransportContext.Key<Map<String, String>> DTLS_AUTHENTICATION = new TransportContext.Key<>(Collections.emptyMap());
     public static final TransportContext.Key<String> DTLS_PEER_CERTIFICATE_SUBJECT = new TransportContext.Key<>(null);
     public static final TransportContext.Key<byte[]> DTLS_CID = new TransportContext.Key<>(null);
     public static final TransportContext.Key<Instant> DTLS_SESSION_START_TIMESTAMP = new TransportContext.Key<>(null);
+    public static final TransportContext.Key<Boolean> DTLS_SESSION_EXPIRATION_HINT = new TransportContext.Key<>(false);
```
Comment (szysas) [inline coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java:37]: Is this misleading name? 
Copilot suggested `DTLS_SESSION_PERSISTENCE_HINT`

File: coap-mbedtls/src/test/java/org/opencoap/transport/mbedtls/MbedtlsNettyTest.java [hunk 2]
```diff
@@ -94,6 +108,7 @@ void beforeAll() throws IOException {
                             }
 
                         })
+                        .get("/ctx-hint", __ -> ok("OK!").context(TransportContext.of(DTLS_SESSION_EXPIRATION_HINT, true)).toFuture())
```
Comment (szysas) [inline coap-mbedtls/src/test/java/org/opencoap/transport/mbedtls/MbedtlsNettyTest.java:111]: This filter, could be converted to its own class, under main scope. We should also require that incoming request is nonconfirmable and fail (4.00) if it's not.

**Review by szysas — CHANGES_REQUESTED** [review]

File: coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java (hunk 1 above)
Comment (akolosov-n) [reply coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java:37]: `DtlsSessionContext` from `kotlin-mbedlts` already has property `sessionExpirationHint`, so this change should go also there.

I'm also not sure if `persistence hint` is less misleading as it doesn't imply (for those who don't know) that the session will not be longer available after persistence operation.

The meaning of `session expiration hint` is `hint the underlying transport layer that the session could be prematurely scheduled for expiration`.

File: coap-core/src/main/java/com/mbed/coap/packet/CoapResponse.java [hunk 3]
```diff
@@ -142,26 +156,27 @@ public String toString() {
...
         private Builder(Code code) {
             this.code = code;
         }
 
         public CoapResponse build() {
-            return new CoapResponse(code, payload, options.build());
+            return new CoapResponse(code, payload, options.build(), transContext);
         }
 
         public SeparateResponse toSeparate(Opaque token, InetSocketAddress peerAddress, TransportContext transContext) {
```
[the remark this answers is not in the document]
Comment (akolosov-n) [reply coap-core/src/main/java/com/mbed/coap/packet/CoapResponse.java:182]: Actually not that a big change if we will keep the current `SeparateResponse` interface and just use `response` transport context internally.

[the remark this answers is not in the document]
Comment (akolosov-n) [reply coap-core/src/main/java/com/mbed/coap/packet/CoapResponse.java:182]: See commit https://github.com/open-coap/java-coap/pull/89/commits/4c95e3a140f1cb30059af58f9a135133a4247f74

File: coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java (hunk 1 above)
Comment (szysas) [reply coap-mbedtls/src/main/java/org/opencoap/transport/mbedtls/DtlsTransportContext.java:37]: OK, lets leave it as now, and when we figure out better name then rename it.

File: coap-core/src/main/java/com/mbed/coap/packet/CoapResponse.java [hunk 4]
```diff
@@ -99,14 +101,23 @@ public String getPayloadString() {
...
     }
 
+    @Deprecated
     public SeparateResponse toSeparate(Opaque token, InetSocketAddress peerAddress, TransportContext transContext) {
-        return new SeparateResponse(this, token, peerAddress, transContext);
+        return new SeparateResponse(this.withContext(transContext), token, peerAddress);
     }
 
     public SeparateResponse toSeparate(Opaque token, InetSocketAddress peerAddress) {
         return toSeparate(token, peerAddress, TransportContext.EMPTY);
```
Comment (szysas) [inline coap-core/src/main/java/com/mbed/coap/packet/CoapResponse.java:110]: If above method is deprecated (which make sense) then this one should use directly constructor.

**Review by szysas — APPROVED** [review]
👍

**Pull request merged**